crates/stages/src/stages/stream.rs

```diff
@@ -0,0 +1,144 @@
+use futures_util::stream::Stream;
+use num_traits::One;
+use std::{
+    cmp::Ordering,
+    collections::{binary_heap::PeekMut, BinaryHeap},
+    ops::Add,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// A Stream type that emits key-value pairs in sequential order of the key.
+#[pin_project::pin_project]
+#[must_use = "stream does nothing unless polled"]
+pub(crate) struct SequentialPairStream<Key: Ord, Value, St> {
+    /// The next item we expect from the stream
+    next: Key,
+    /// buffered entries
+    pending: BinaryHeap<OrderedItem<Key, Value>>,
+    #[pin]
+    stream: St,
+    done: bool,
+}
+
+// === impl SequentialPairStream ===
+
+impl<Key: Ord, Value, St> SequentialPairStream<Key, Value, St> {
```

Comment:
Should prob be in `common` somewhere? I can imagine we might want to re-use somewhere else (downloaders?)

Reply:
if there's no actual benefit from using `append` over `insert` here then we perhaps don't even need this but, yes should be moved to common

Reply:
there should be a benefit over larger sets of data since `insert` performs a search - for a small set of data, this is probably ok (searches terminate fast since it's ordered) but for larger sets it will become an issue I think